src/Compiler/Facilities/AsyncMemoize.fs

```diff
@@ -153,409 +141,108 @@ type internal AsyncMemoize<'TKey, 'TVersion, 'TValue when 'TKey: equality and 'T
+            }
 
-            match!
-                processRequest post (key, GetOrCompute(computation, ct)) callerDiagnosticLogger
-                |> Async.AwaitTask
-            with
-            | New internalCt ->
+        let getOrAdd () =
+            let cached, otherVersions = cache.GetAll(key.Key, key.Version)
 
-                let linkedCtSource = CancellationTokenSource.CreateLinkedTokenSource(ct, internalCt)
-                let cachingLogger = new CachingDiagnosticsLogger(Some callerDiagnosticLogger)
+            let countHit v = Interlocked.Increment &hits |> ignore; v
+            let cacheSetNewJob () =
+                let job = Job(wrappedComputation, cancelUnawaited = cancelUnawaitedJobs)
+                cache.Set(key.Key, key.Version, key.Label, job)
+                job
 
-                try
-                    return!
-                        Async.StartAsTask(
-                            async {
-                                // TODO: Should unify starting and restarting
-                                let currentLogger = DiagnosticsThreadStatics.DiagnosticsLogger
-                                DiagnosticsThreadStatics.DiagnosticsLogger <- cachingLogger
-
-                                log (Started, key)
-
-                                try
-                                    let! result = computation
-                                    post (key, (JobCompleted(result, cachingLogger.CapturedDiagnostics)))
-                                    return result
-                                finally
-                                    DiagnosticsThreadStatics.DiagnosticsLogger <- currentLogger
-                            },
-                            cancellationToken = linkedCtSource.Token
-                        )
-                        |> Async.AwaitTask
-                with
-                | TaskCancelled ex ->
-                    // TODO: do we need to do anything else here? Presumably it should be done by the registration on
-                    // the cancellation token or before we triggered our own cancellation
+            otherVersions,
 
-                    // Let's send this again just in case. It seems sometimes it's not triggered from the registration?
+            cached
+            |> Option.map countHit
+            |> Option.defaultWith cacheSetNewJob
 
-                    Interlocked.Increment &cancel_exception_original |> ignore
+        async {            
+            let otherVersions, job = lock cache getOrAdd
 
-                    post (key, (OriginatorCanceled))
-                    return raise ex
-                | ex ->
-                    post (key, (JobFailed(ex, cachingLogger.CapturedDiagnostics)))
-                    return raise ex
+            log Requested key
+
+            if cancelDuplicateRunningJobs && not cancelUnawaitedJobs then
+                otherVersions |> Seq.map snd |> Seq.iter _.CancelIfUnawaited()
```

Comment:
I think the original idea was to cancel them even if they're still awaited. But we can also decide not to support that and clean it up further.

Reply:
Yep, currently this code does nothing, because all the caches in use have the default `cancelUnawaitedJobs` true. 
I thought about it a lot. The thing is, you cannot really cancel a async request without access to the original cts. The awaiting client would just receive an exception instead of a result?

Reply:
Yeah. If someone is asking us to e.g. check a file and then again a new version of the same file, without cancelling the original request, they would get an exception on that. Because we can say that original request is outdated. But it really should be the responsibility of the caller, so we can just not deal with it here.